nixos/modules/security/doas.nix

```diff
@@ -0,0 +1,259 @@
+            };
+
+            groups = mkOption {
+              type = with types; listOf (either str int);
+              default = [];
+              description = "The groups / GIDs this rule should apply for.";
+            };
+
+            runAs = mkOption {
+              type = with types; nullOr str;
+              default = null;
+              description = ''
+                Under which user/group the specified command is allowed to run.
+
+                A user can be specified using just the username: <code>"foo"</code>. It is also
+                possible to only allow running as a specific group with <code>":bar"</code>.
+              '';
+            };
+
+            cmd = mkOption {
+              type = with types; nullOr str;
+              default = null;
+              description = ''
+                The command the user is allowed to run. The default is all commands. Be advised that
+                it is best to specify absolute paths. If a relative path is specified, only a
+                restricted PATH will be searched. When set to <code>null</code>, all commands are
+                allowed.
+              '';
+            };
+
+            args = mkOption {
+              type = with types; nullOr (listOf str);
+              default = [];
+              description = ''
+                Arguments to the command. The arguments provided by the user need to match those
+                specified. When set to <code>null</code>, the command must be run without any
+                arguments.
+              '';
+            };
+          };
+        }
+      );
+    };
+
+    extraConfig = mkOption {
+      type = with types; lines;
+      default = "";
+      description = ''
+        Extra configuration text appended to <filename>doas.conf</filename>.
+      '';
+    };
+  };
+
+
+  ###### implementation
+
+  config = mkIf cfg.enable {
+
+    security.doas.extraRules = [
+      {
+        groups = [ "wheel" ];
+        noPass = !cfg.wheelNeedsPassword;
+      }
+    ];
+
+    security.wrappers = {
+      doas.source = "${doas}/bin/doas";
+    };
+
+    environment.systemPackages = [
+      doas
+    ];
+
+    security.pam.services.doas = {
+      allowNullPassword = true;
+      sshAgentAuth = true;
+    };
+
+    environment.etc."doas.conf" = {
+      source = pkgs.runCommand "doas-conf"
+        {
+          src = pkgs.writeText "doas-conf-in" ''
+            # Do NOT edit this file by hand. Set the NixOS options
+            # ‘security.doas.extraRules’ or ‘security.doas.extraConfig’ instead.
+            # In order to completely replace the contents of this file, use
+            # ‘environment.etc."doas.conf"’.
+
+            # "root" is allowed to do anything.
+            permit nopass keepenv root
+
+            # extraRules
+            ${concatStringsSep "\n" (
+            lists.flatten (
+              map (
+                rule:
```

Comment:
This is a very big thing to have inside a string interpolation. I think it should be moved to a let binding at the top of the file.

Reply:
I definitely like it better in its new home, but maybe it could also be broken up into smaller functions? It’s very big for a single function and feels a bit intimidating to read. Of course, there’s a chance breaking it up could make it worse, but I think as long as the functions were well-named it would be okay. What do you think?

Reply:
I've split out the `opts` and `args` stuff in 15f5195f492d6118ca6f140dc4e339f874398ecc. They are now `mkOpts` and `mkArgs`, respectively. I didn't want to try and split out the actual `map`ping of `rule.users`/`rule.groups` due to the `let`-bound references.

How'd I do?

(Also: lol my fixups are getting out of hand)